### PR TITLE
added IsNullFixer: transforms is_null(...) -> null === ...

### PR DESCRIPTION
**Pull request by kalessil**
!is_null will be transformed as null !==.

### CODE REVIEW

**Comment by kalessil** [comment]
@keradus : as I promised, is_null fixer provided

File: Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            array('<?php $x = new ScopeB\\is_null(trim($x));'),
+
+            array('<?php is_nullSmth(trim($x));'),
+            array('<?php smth_is_null(trim($x));'),
+
+            array('<?php "SELECT ... is_null(trim($x)) ...";'),
+            array('<?php "SELECT ... is_null(trim($x)) ...";'),
+            array('<?php "test" . "is_null" . "in concatenation";'),
+
+            array('<?php $x = null === trim($x);', '<?php $x = is_null(trim($x));'),
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: These tests are a bit of a bad example since trim never returns null.

Comment (kalessil) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: Updated

File: Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public function provideExamples()
+    {
+        $multiLinePatternToFix = <<<FIX
+<?php \$x =
+is_null
+
+(
+    json_decode
+    (
+        \$x
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: use nowdoc and you won't need to escape that

File: Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            array('<?php $x = null !== json_decode($x);', '<?php $x = ! \\is_null(json_decode($x));'),
+            array('<?php $x = null !== json_decode($x);', '<?php $x = ! \\is_null( json_decode($x) );'),
+
+            array('<?php $x = null === json_decode($x).".dist";', '<?php $x = is_null(json_decode($x)).".dist";'),
+            array('<?php $x = null !== json_decode($x).".dist";', '<?php $x = !is_null(json_decode($x)).".dist";'),
+            array('<?php $x = null === json_decode($x).".dist";', '<?php $x = \\is_null(json_decode($x)).".dist";'),
+            array('<?php $x = null !== json_decode($x).".dist";', '<?php $x = !\\is_null(json_decode($x)).".dist";'),
+
+            array($multiLinePatternFixed, $multiLinePatternToFix),
+            array('<?php $x = /**/null === /**//** */json_decode($x)/***//*xx*/;', '<?php $x = /**/is_null/**/ /** x*/(/**//** */json_decode($x)/***/)/*xx*/;'),
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: no comments must be lost

Comment (kalessil) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: Okay, logic change needed as well

**Comment by keradus** [comment]
Great to see that!

It would be great to have having some recursive calls like `is_null(is_null($foo) ? is_null($bar) : baz(is_null($baz)))`

File: Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php (hunk 2 above)
Comment (kalessil) [inline Symfony/CS/Tests/Fixer/Contrib/IsNullFixerTest.php]: Thanks for the hint

**Comment by kalessil** [comment]
Can be done easily - will add UTs as well for this case.

**Comment by kalessil** [comment]
@keradus : is_null fixer behavior corrected

**Comment by gharlan** [comment]
Could be a candidate for symfony level by golden rule: https://github.com/symfony/symfony/search?utf8=✓&q=is_null

**Comment by keradus** [comment]
That's great!

**Comment by gharlan** [comment]
```php
if (!is_null($a = foo()) && $a->bar()) {}
```

I guess the "fixed" result is:

```php
if (null !== $a = foo() && $a->bar()) {}
```

But this is not the same.

**Comment by stof** [comment]
@gharlan are you sure ?

**Comment by gharlan** [comment]
yes. The result is the same as:

```php
if (null !== ($a = foo() && $a->bar())) {}
```

http://3v4l.org/ElKmh

**Comment by stof** [comment]
then you would have to keep the parenthesis around the expression inside ``is_null`` (which makes the code even simpler btw)

**Comment by kalessil** [comment]
Closing, as we've decided to merge this into #1307

**Comment by keradus** [comment]
@kalessil #1308 didn't address `is_null` => `null ===` in the end.
Could you please finish this PR? it would be so great !

**Comment by kalessil** [comment]
Checking this PR ATM: need to recall where I've stopped...

**Comment by kalessil** [comment]
Can anyone recommend a Vagrant configuration for Ubuntu/PHP 5.3 dev-environment for finishing this?

**Comment by kalessil** [comment]
All, please follow #2415 instead of this

**Pull request closed without merging**